src/systems/systemDialog.component.tsx

```diff
@@ -14,110 +18,109 @@ import {
 
-const getEmptySystem = (): AddSystem => {
-  return {
-    // Here using null for optional values only, so that types for isUpdated parameters
-    // can match
-    name: '',
-    description: null,
-    location: null,
-    owner: null,
-    importance: SystemImportanceType.MEDIUM,
-  } as AddSystem;
-};
+const systemsSchema = z.object({
+  name: z.string().trim().min(1, { message: 'Please enter a name' }),
+  location: z
+    .string()
+    .optional()
+    .transform((val) => (val === '' ? null : val)),
+  owner: z
+    .string()
+    .optional()
+    .transform((val) => (val === '' ? null : val)),
+  description: z
+    .string()
+    .optional()
+    .transform((val) => (val === '' ? null : val)),
+  importance: z.nativeEnum(SystemImportanceType),
+});
 
+export type SystemDialogType = 'add' | 'edit' | 'save as';
 export interface SystemDialogProps {
   open: boolean;
   onClose: () => void;
-  type: SystemDialogType;
-  // Only required for add
+  type: 'add' | 'edit' | 'save as';
   parentId?: string | null;
-  // Only required for prepopulating fields for an edit dialog
   selectedSystem?: System;
 }
 
-const SystemDialog = React.memo((props: SystemDialogProps) => {
+function SystemDialog(props: SystemDialogProps) {
   const { open, onClose, parentId, type, selectedSystem } = props;
-
-  // User entered properties
-  const [systemData, setSystemData] =
-    React.useState<AddSystem>(getEmptySystem());
-
-  // Ensure system data is updated when the selected system changes
-  useEffect(() => {
-    if (open) {
-      if (type === 'add') setSystemData(getEmptySystem());
-      else if (selectedSystem) setSystemData(selectedSystem as AddSystem);
-    }
-  }, [selectedSystem, open, type]);
-
   // Error messages for the above properties (undefined means no error)
   const [nameError, setNameError] = React.useState<string | undefined>(
     undefined
   );
 
-  // Form error that should dissappear when the form is modified
+  // Form error that should disappear when the form is modified
   const [formError, setFormError] = React.useState<string | undefined>(
     undefined
   );
 
-  const handleClose = React.useCallback(() => {
-    if (type === 'add') setSystemData(getEmptySystem());
-    // Reset for edit
-    else setSystemData(selectedSystem as AddSystem);
-
-    // Remove all errors
-    setNameError(undefined);
-    setFormError(undefined);
+  const isNotAdding = type !== 'add' && selectedSystem;
+  const system = selectedSystem || {
+    name: '',
+    description: '',
+    location: '',
+    owner: '',
+    importance: SystemImportanceType.MEDIUM,
+  };
 
-    onClose();
-  }, [onClose, selectedSystem, type]);
+  const initialSystem: AddSystem = {
+    name: isNotAdding ? system.name : '',
+    description: isNotAdding ? system.description : '',
+    location: isNotAdding ? system.location : '',
+    owner: isNotAdding ? system.owner : '',
+    importance: system.importance,
+  };
```

Comment:
In this case I think you can use something like
  ```const initialSystem: AddSystem = type !== 'add' && selectedSystem
    ? (selectedSystem as AddSystem)
    : {
        name: '',
        description: '',
        location: '',
        owner: '',
        importance: SystemImportanceType.MEDIUM,
      };
```
To avoid repetition - It looks like the zod schema just ignores the extra fields.